remill/BC/DeadStoreEliminator.cpp

```diff
@@ -0,0 +1,532 @@
+        return VisitResult::NoProgress;
+      } else {
+        auto offset = (plus ? ptr->second + cint->getZExtValue() : ptr->second - cint->getZExtValue());
+        offset_map.insert({&I, offset});
+        LOG(INFO) << "offsetting: " << LLVMThingToString(&I);
+        return VisitResult::Progress;
+      }
+    } else {
+      // check if both are in the offset_map
+      auto ptr1 = offset_map.find(val1);
+      auto ptr2 = offset_map.find(val2);
+      if (ptr1 == offset_map.end() || ptr2 == offset_map.end()) {
+        return VisitResult::NoProgress;
+      } else {
+        auto offset = (plus ? ptr1->second + ptr2->second : ptr1->second - ptr2->second);
+        offset_map.insert({&I, offset});
+        LOG(INFO) << "offsetting: " << LLVMThingToString(&I);
+        return VisitResult::Progress;
+      }
+    }
+  }
+}
+
+// Visit a PHI node and update the offset map.
+VisitResult ForwardAliasVisitor::visitPHINode(llvm::PHINode &I) {
+  LOG(INFO) << "Entered PHI node";
+  // iterate over each operand
+  auto in_offset_map = false;
+  uint64_t offset;
+  for (auto &operand : I.operands()) {
+    if (exclude.count(operand)) {
+      // fail if some operands are excluded and others are state offsets
+      if (in_offset_map) {
+        return VisitResult::Error;
+      }
+    } else {
+      auto ptr = offset_map.find(operand);
+      if (ptr == offset_map.end()) {
+        return VisitResult::NoProgress;
+      } else {
+        if (!in_offset_map) {
+          offset = ptr->second;
+          in_offset_map = true;
+        } else {
+          if (ptr->second != offset) {
+            // bail if the offsets don't match
+            return VisitResult::Error;
+          }
+        }
+      }
+    }
+  }
+  if (in_offset_map) {
+    LOG(INFO) << "offsetting: " << LLVMThingToString(&I);
+    offset_map.insert({&I, offset});
+  } else {
+    LOG(INFO) << "excluding: " << LLVMThingToString(&I);
+    exclude.insert(&I);
+  }
+  return VisitResult::Progress;
+}
+
+// Return the scope of the given instruction.
+llvm::MDNode *GetScopeFromInst(llvm::Instruction &I) {
+  llvm::AAMDNodes N;
+  I.getAAMetadata(N);
+  return N.Scope;
+}
+
+// Remove all dead stores.
+void removeDeadStores(std::unordered_set<llvm::Instruction *> dead_stores) {
+  /* for (auto *store : dead_stores) { */
+  /*   store->eraseFromParent(); */
+  /* } */
+}
+
+LiveSetBlockVisitor::LiveSetBlockVisitor(std::unordered_map<llvm::MDNode *, uint64_t> scope_to_offset_)
+  : scope_to_offset(scope_to_offset_), curr_wl(), next_wl(), block_map(), to_remove(), live() { }
+
+void LiveSetBlockVisitor::addFunction(llvm::Function &func) {
+  for (auto &block : func) {
+    // The machines rose from the ashes of the nuclear fire....
+    if (block.getTerminator()) {
+      curr_wl.push_back(&block);
+      // add block to map with initial entry and exit sets
+      block_map.insert({&block, std::make_pair(LiveSet(), LiveSet())});
+    }
+  }
+}
+
+void LiveSetBlockVisitor::visit() {
```

Comment:
`void` inside of empty parens. If you have a function that takes no arguments, always put in `void`.

Reply:
Rename to `Visit`